pkg/topolvm-scheduler/cmd/root.go

```diff
@@ -80,23 +86,26 @@ func subMain() error {
 		return err
 	}
 
-	serv := &well.HTTPServer{
-		Server: &http.Server{
-			Addr:    config.ListenAddr,
-			Handler: h,
-		},
+	serv := &http.Server{
+		Addr:        config.ListenAddr,
+		Handler:     accessLogHandler(parentCtx, h),
+		ReadTimeout: 30 * time.Second,
 	}
 
-	err = serv.ListenAndServe()
-	if err != nil {
-		return err
-	}
-	err = well.Wait()
+	ctx, stop := signal.NotifyContext(parentCtx, os.Interrupt, syscall.SIGTERM)
+	defer stop()
+
+	go func() {
+		<-ctx.Done()
+		if err := serv.Shutdown(parentCtx); err != nil {
+			logger.Error(err, "failed to shutdown gracefully")
+		}
+	}()
 
-	if err != nil && !well.IsSignaled(err) {
+	err = serv.ListenAndServe()
+	if !errors.Is(err, http.ErrServerClosed) {
```

Comment:
Unlike `Server.Serve()` in grpc-go, `Server.ListenAndServe()` in net/http  does not wait until all connections are closed even if `Server.Shutdown()` is called. Instead, `ListenAndServe()` returns immediately, so we need to wait until `Shutdown()` returns to shutdown gracefully. With this code the main go routine exits when `Shutdown()` called, which results in the process itself finishes immediately without waiting existing connections. Is this ok?

> When Shutdown is called, Serve, ListenAndServe, and ListenAndServeTLS immediately return ErrServerClosed. Make sure the program doesn't exit and waits instead for Shutdown to return. 

https://pkg.go.dev/net/http#Server.Shutdown

Reply:
It seems not ok to me. Thank you for pointing this out. I'll fix my code.

Reply:
Now it's fixed. The change I made: https://github.com/topolvm/topolvm/compare/0684eb256a5fb0afddcc47124b79fc3a0757413a..6c6bf8ca1a3f1bb8c6d93d365505386543fa7103

Reply:
Perfect! Thank you.